webui/react/src/pages/ProjectDetails.tsx

```diff
@@ -167,42 +168,55 @@ const ProjectDetails: React.FC = () => {
       const states = (settings.state || []).map((state) => (
         encodeExperimentState(state as RunState)
       ));
-      const response = await getExperiments(
-        {
-          archived: settings.archived ? undefined : false,
-          labels: settings.label,
-          limit: settings.tableLimit,
-          name: settings.search,
-          offset: settings.tableOffset,
-          orderBy: settings.sortDesc ? 'ORDER_BY_DESC' : 'ORDER_BY_ASC',
-          projectId: id,
-          sortBy: validateDetApiEnum(V1GetExperimentsRequestSortBy, settings.sortKey),
-          states: validateDetApiEnumList(Determinedexperimentv1State, states),
-          users: settings.user,
-        },
-        { signal: canceler.signal },
+      const baseParams: GetExperimentsParams = {
+        archived: settings.archived ? undefined : false,
+        labels: settings.label,
+        name: settings.search,
+        orderBy: settings.sortDesc ? 'ORDER_BY_DESC' : 'ORDER_BY_ASC',
+        projectId: id,
+        sortBy: validateDetApiEnum(V1GetExperimentsRequestSortBy, settings.sortKey),
+        states: validateDetApiEnumList(Determinedexperimentv1State, states),
+        users: settings.user,
+      };
+      const pinnedIds = (settings.pinned[id] ?? []);
+      const pinnedExperimentsRequest = getExperiments({
+        ...baseParams,
+        experimentFilter: { experimentIds: pinnedIds, restriction: 'RESTRICTION_INCLUDE' },
+        limit: settings.tableLimit,
+        offset: 0,
+      }, { signal: canceler.signal });
+      const otherExperimentsRequest = getExperiments({
+        ...baseParams,
+        experimentFilter: { experimentIds: pinnedIds, restriction: 'RESTRICTION_EXCLUDE' },
+        limit: settings.tableLimit - pinnedIds.length,
+        offset: settings.tableOffset -
+            (settings.tableOffset / settings.tableLimit) * pinnedIds.length,
+      }, { signal: canceler.signal });
+      const allRequests = [ pinnedExperimentsRequest, otherExperimentsRequest ];
+      const [ pinnedExpResponse, otherExpResponse ] = await Promise.all(allRequests);
+      setTotal(
+        (pinnedExpResponse.pagination.total ?? 0) + (otherExpResponse.pagination.total ?? 0),
       );
-      setTotal(response.pagination.total ?? 0);
-      setExperiments((prev) => {
-        if (isEqual(prev, response.experiments)) return prev;
-        return response.experiments;
-      });
+      setExperiments([ ...pinnedExpResponse.experiments, ...otherExpResponse.experiments ]);
```

Comment:
fix: There's some sort of race condition when it comes to pinning/unpinning and switching pages

https://user-images.githubusercontent.com/15078396/189415696-dec0a744-38df-492c-905b-ea6d062d9de6.mov

Reply:
I think its `useSettings` bug.  In DB, it is stored as expected but `useSettings` looks at the old data